### DISCUSSION TITLE
How to set a constant buoyancy field

### DISCUSSION THREAD

**Discussion by RWBY-wiess** [CliMA/Oceananigans.jl#2683] [Computational science!]
I want to set a constant buoyancy field, in other words, it doesn't change over time. But I don't know what should I do, is there any way?
In addition, except for a constant buoyancy field, measures that can set a constant pressure field or a constant external force (the right-hand term of the momentum equation) are also helpful.

**Comment by glwagner** [comment]
Constant buoyancy fields and pressure fields can be formulated as forcing terms. Check out the documentation for `Forcing` here:

https://clima.github.io/OceananigansDocumentation/stable/model_setup/forcing_functions/

**Reply by RWBY-wiess** [reply]
Thanks for your advice. Unluckily, I seem to have some trouble with using forcing .
This is part of my code. After running for 200 seconds, I can see that the buoyancy field is getting bigger and bigger with time.

`using Oceananigans
using Oceananigans.TurbulenceClosures
using Printf

no_slip_bc = ValueBoundaryCondition(0.0)
no_slip = FieldBoundaryConditions(no_slip_bc)
BC1 = FieldBoundaryConditions(no_slip_bc, top = GradientBoundaryCondition(0), bottom = GradientBoundaryCondition(0))
grid = RectilinearGrid(size=(128, 128), x=(-20000, 20000), z=(-300, 0), halo = (3, 3), topology = (Bounded, Flat, Bounded))

horizontal_closure = HorizontalScalarDiffusivity(ν = 0)
vertical_closure = VerticalScalarDiffusivity(ν = 4e-2)

coriolis = FPlane(latitude=30)

B(x, y, z, t) = - tanh(x/4000) * exp(z/30) * (1e-3)

model = NonhydrostaticModel(grid = grid,
advection = CenteredSecondOrder(),
boundary_conditions = (u = BC1, v = BC1),
coriolis = coriolis,
closure=(horizontal_closure, vertical_closure),
tracers = (:b),
buoyancy = BuoyancyTracer(),
forcing = (b = B,))

model.forcing.b`

**Reply by simone-silvestri** [reply]
Yes, if you use a forcing you are introducing a term in the RHS of the specific equation. This means that in your case (since you force buoyancy)
$$\frac{\partial b}{\partial t} = G_b + F_b$$

if the forcing is constant it will always provide a source/sink for your field that will add up in time, so probably you don't want to add a forcing to the buoyancy field. 

Depending on what you want to do you have two options:
(1) restore your buoyancy field to a constant field. In this case the forcing will be dependent on the buoyancy field $$F_b = \lambda * (b_{target} - b)$$ where b is your buoyancy tracer and lambda a restoring coefficient. (you can do it by passing a function to the `Forcing` constructor, you can find an example in `validation/mesoscale_turbulence/abernathy_channel.jl`). In this case your buoyancy field will not be _formally_ constant but it will oscillate around your prescribed values
(2) Forget about buoyancy (`buoyancy = nothing` in the model constructor) and directly force velocities with your field of choice. In this case you will have
$$\frac{\partial \boldsymbol{u}}{\partial t} = G_{\boldsymbol{u}} + F_{\boldsymbol{u}}$$ and you can specify the forcing to mimic the effect of a constant buoyancy/pressure field

**Reply by glwagner** [reply]
`BackgroundField` may also be relevant: https://clima.github.io/OceananigansDocumentation/stable/model_setup/background_fields/

A description of the equations solved by `NonhydrostaticModel` is here: https://clima.github.io/OceananigansDocumentation/stable/physics/nonhydrostatic_model/

If you can, it often helps to start from an example or validation test, if one exists that is similar enough to the experiment you are trying to implement.
